packages/components/src/ui/context/context-system-provider.js

```diff
@@ -10,52 +10,68 @@ import {
 	createContext,
 	useContext,
 	useRef,
-	useState,
+	useEffect,
+	useMemo,
 	memo,
 } from '@wordpress/element';
-
-/**
- * Internal dependencies
- */
-import { useIsomorphicLayoutEffect } from '../utils';
+import warn from '@wordpress/warning';
 
 export const ComponentsContext = createContext(
 	/** @type {Record<string, any>} */ ( {} )
 );
 export const useComponentsContext = () => useContext( ComponentsContext );
 
+/**
+ * Runs an effect only on update (i.e., ignores the first render)
+ *
+ * @param {import('react').EffectCallback} effect
+ * @param {import('react').DependencyList} deps
+ */
+function useUpdateEffect( effect, deps ) {
+	const mounted = useRef( false );
+	useEffect( () => {
+		if ( mounted.current ) {
+			return effect();
+		}
+		mounted.current = true;
+		return undefined;
+	}, deps );
+}
+
 /**
  * Consolidates incoming ContextSystem values with a (potential) parent ContextSystem value.
  *
+ * Note: This function will warn if it detects an un-memoized `value`
+ *
  * @param {Object}              props
  * @param {Record<string, any>} props.value
  * @return {Record<string, any>} The consolidated value.
  */
 function useContextSystemBridge( { value } ) {
 	const parentContext = useComponentsContext();
-	const parentContextRef = useRef( parentContext );
-	const valueRef = useRef( merge( parentContext, value ) );
-
-	const [ config, setConfig ] = useState( valueRef.current );
 
-	useIsomorphicLayoutEffect( () => {
-		let hasChange = false;
+	const valueRef = useRef( value );
 
-		if ( ! isEqual( value, valueRef.current ) ) {
-			valueRef.current = value;
```

Comment:
It turns out this was the problematic line that was causing the context to not get merged correctly. Becuase `valueRef` was being assigned to the merged value, value and valueRef.current were never going to be equal if the parent context included anything.

We're able to avoid this comparison altogether if we make a guarantee that value and parent context are memoized however, so the new version of this hook is able to avoid this complexity completely.